### PR TITLE
Support distributed Adam with T5 and support overlapped grad reductions with pipeline parallelism

### PR DESCRIPTION
**Pull request by timmoon10**
# What does this PR do ?

Generalize distributed Adam support for GPT-3 to T5 and other Megatron-LM models. It also implements several performance optimizations.

**Collection**: NLP
# Changelog 
- When params are BF16, distributed Adam will store 16-bit param remainders instead of FP32 main params
- Decouples distributed Adam support from Megatron O2-level optizations
- Add support for Apex distributed Adam optimizer with other Megatron-LM models, namely T5
- Add support for overlapped grad reductions with pipeline or sequence parallelism

# Usage

Set optimizer to `distributed_fused_adam` in config file:

https://github.com/NVIDIA/NeMo/blob/265f7b17876aef7937cf0fedf7a76885bdc63288/examples/nlp/language_modeling/conf/megatron_t5_config.yaml#L137

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Closes https://github.com/NVIDIA/NeMo/pull/4799
* Depends on https://github.com/NVIDIA/apex/pull/1475

### CODE REVIEW

**Comment by timmoon10** [comment]
Running T5 41B on 32 Selene nodes, I see a 1.2x speedup over the pure data-parallel impl, 66% of expected memory savings, and nearly identical loss values after 20 steps.

[Full results with T5 41B and GPT-3 175B](https://docs.google.com/spreadsheets/d/1wqnJ0l1Yuogz-XElzaNLIG4Geo_UhCAwdwkdiBWGrkU/edit#gid=525117936). The run configurations are detailed inside. Note that I ran with a relatively small global batch size, which makes the communication cost more significant.

**Comment by timmoon10** [comment]
I've fixed a bug related to gradient scaling that only affected FP16 training (see https://github.com/NVIDIA/apex/pull/1512). BF16 training shouldn't be affected.

This PR is ready to merge. Pinging @erhoo82.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -254,6 +307,27 @@ def training_step(self, batch, batch_idx):
...
+                and self.cfg.get('pipeline_model_parallel_size', 1) == 1
+                and not self.cfg.get('sequence_parallel', False)
+            ):
+                custom_sync_context_handler = self._optimizer.no_sync
+            else:
+                # TODO: enable async grad all reduce with O1/autocast
+                # mixed precision training, with pipeline parallelism,
+                # or with sequence parallelism
+                custom_sync_context_handler = None
+
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:330]: Can we move this to the base model? GPT has the same code.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -236,16 +260,28 @@ def training_step(self, batch, batch_idx):
...
+                # keep grad tensors around
+                custom_sync_context_handler = lambda: self._optimizer.no_sync(greedy_grad_copy=False)
         else:
-            # TODO: enable async grad all reduce for O1/autocast mixed precision training
-            custom_sync_context_handler = None
+            if self.megatron_amp_o2 and not self.cfg.get('sequence_parallel', False):
+                custom_sync_context_handler = self._optimizer.no_sync
+            else:
+                # TODO: enable async grad all reduce for O1/autocast mixed precision training
+                custom_sync_context_handler = None
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:276]: Can we move this to the base model? T5 has the same code.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 3]
```diff
@@ -293,7 +362,15 @@ def training_step(self, batch, batch_idx):
...
-        if self.megatron_amp_o2:
+        if self.with_distributed_adam:
+            # launch grad reductions
+            # Note: grads in first pipeline stage have already been
+            # reduced
+            if not parallel_state.is_pipeline_first_stage():
+                self._optimizer.try_grad_sync(
+                    p for p in self.parameters() if not getattr(p, '_disable_overlap_grad_sync', False)
+                )
+        elif self.megatron_amp_o2:
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:373]: Can we move this to the base model? The same code is in GPT.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 4]
```diff
@@ -277,8 +313,13 @@ def training_step(self, batch, batch_idx):
...
         if self.with_distributed_adam:
-            # gradients are reduced internally in distributed optimizer
-            pass
+            # launch grad reductions
+            # Note: grads in first pipeline stage have already been
+            # reduced
+            if not parallel_state.is_pipeline_first_stage():
+                self._optimizer.try_grad_sync(
+                    p for p in self._optimizer.parameters() if not getattr(p, '_disable_overlap_grad_sync', False)
+                )
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:322]: Can we move this to the base model? The same code is in T5.

File: nemo/core/optim/distributed_adam.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ericharper) [inline nemo/core/optim/distributed_adam.py:1]: 2022

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 1 above)
Comment (timmoon10) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:330]: In the `main` branch, the non-distopt behavior with [GPT](https://github.com/NVIDIA/NeMo/blob/86565741f22d0c19da07680e2f40bc207c613206/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L241) and [T5](https://github.com/NVIDIA/NeMo/blob/86565741f22d0c19da07680e2f40bc207c613206/nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py#L257) is subtly different. In particular, T5 doesn't pass in a context if pipeline parallelism is enabled. We could make the behavior between GPT and T5 consistent, but I think that would be better in a standalone PR so it's easier to identify any regressions.

**Comment by timmoon10** [comment]
Running on a DGX A100 node for 50 steps with 2-way data, tensor, and pipeline parallelism, I see nearly identical learning behavior with and without the distributed optimizer:

| Model | ZeRO | O2 | Data type | Throughput | Train loss | Val loss |
|---|---|---|---|---|---|---|
| GPT-2 124M | Yes | No | FP16 |  2.69it/s | 8.4 | 7.870 |                                                                                                               
| GPT-2 124M | No | No | FP16 |  3.31it/s | 8.4 | 7.870 |                                                                                                                
| GPT-2 124M | Yes | No | BF16 |  3.19it/s | 8.28 | 7.790 |                                                                                                              
| GPT-2 124M | No | No | BF16 |  3.39it/s | 8.28 | 7.790 |                                                                                                               
| GPT-2 124M | Yes | Yes | BF16 |  3.44it/s | 8.3 | 7.800 |
| GPT-2 124M | No | Yes | BF16 |  3.60it/s | 8.3 | 7.800 |
| T5 220M | Yes | No | FP32 |  1.46it/s | 7.64 | 7.530 |                                                                                                                
| T5 220M | No | No | FP32 |  1.69it/s | 7.64 | 7.530 |                                                                                                                  
| T5 220M | Yes | No | FP16 |  1.43it/s | 8.45 | 8.290 |                                                                                                                
| T5 220M | No | No | FP16 |  1.43it/s | 8.45 | 8.290 |                                                                                                                  
| T5 220M | Yes | No | BF16 |  1.50it/s | 7.66 | 7.560 |                                                                                                                 
| T5 220M | No | No | BF16 |  1.45it/s | 7.65 | 7.550 |                                                                                                                  
| T5 220M | Yes | Yes | BF16 |  1.58it/s | 7.65 | 7.540 |
| T5 220M | No | Yes | BF16 |  1.61it/s | 7.65 | 7.540 |

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Review by ericharper — APPROVED** [review]
Re-approving

**Comment by timmoon10** [comment]
With https://github.com/NVIDIA/apex/pull/1514 the distributed optimizer supports interleaved pipeline parallelism. Running GPT-2 124M for 20 steps, I get the same loss values with and without the distributed optimizer.
